### PR TITLE
En_Dns OK

### PR DESCRIPTION
**Pull request by EllipticEllipsis**
All functions matching, several named. I left unnamed the really short ones that just do purchase checks, and the string of action functions near the end where the purposes are somewhat opaque.

### CODE REVIEW

File: src/overlays/actors/ovl_En_Dns/z_en_dns.h [hunk 1]
```diff
@@ -6,9 +6,34 @@
...
+    /* 0x08 */ EnDnsPurchaseableCheck purchaseableCheck;
+    /* 0x0C */ EnDnsSetRupeesAndFlags setRupeesAndFlags;
+} DnsItemEntry; // size = 0x10
+
 typedef struct EnDns {
     /* 0x0000 */ Actor actor;
-    /* 0x014C */ char unk_14C[0x17C];
+    /* 0x014C */ SkelAnime skelAnime;
+    /* 0x0190 */ Vec3s limbDrawTable[0x12];
+    /* 0x01FC */ Vec3s transitionDrawTable[0x12];
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.h:27]: limb count can be in decimal

File: src/overlays/actors/ovl_En_Dns/z_en_dns.h [hunk 2]
```diff
@@ -6,9 +6,34 @@
...
+    /* 0x0268 */  EnDnsActionFunc actionFunc;
+    /* 0x026C */ ColliderCylinder collider;
+    /* 0x02B8 */ s16 dustTimer;
+    /* 0x02BA */ u8 unk_2BA;
+    /* 0x02BB */ u8 maintainCollider;
+    /* 0x02BC */ u8 standOnGround; // Falls for Burrow when unset
+    /* 0x02BD */ u8 dropCollectible;
+    /* 0x02C0 */ DnsItemEntry* dnsItemEntry; // word-aligned
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.h:35]: both comments unnecessary imo

File: src/overlays/actors/ovl_En_Dns/z_en_dns.h [hunk 3]
```diff
@@ -6,9 +6,34 @@
 
 struct EnDns;
 
+typedef void (*EnDnsActionFunc)(struct EnDns*, GlobalContext*);
+
+typedef u32 (*EnDnsPurchaseableCheck)(struct EnDns*);
+
+typedef void (*EnDnsSetRupeesAndFlags)(struct EnDns*);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.h:13]: id remove the linebreaks

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 4]
```diff
@@ -9,7 +15,36 @@ void EnDns_Destroy(Actor* thisx, GlobalContext* globalCtx);
 void EnDns_Update(Actor* thisx, GlobalContext* globalCtx);
 void EnDns_Draw(Actor* thisx, GlobalContext* globalCtx);
 
-/*
+void EnDns_ChangeAnim(EnDns* this, u8 arg1);
+
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:19]: this forward declaration shouldnt be needed, the function is defined above its first usage

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 5]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+    { 18, 32, 0, { 0, 0, 0 } },
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dns/func_809EFA28.s")
+u16 D_809F040C[] = {
+    0x10A0,
+    // "[sound 3882]I surrender! In return, I will sell  you Deku Nuts! 5 pieces 20 Rupees they are![goto 10A3]"
+    0x10A1,
+    // "[sound 3882]All right! You win! In return, I will sell you Deku Sticks! 1 piece 15 Rupees they are![goto 10A3]"
+    0x10A2,
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:71]: we used to put text box comments but decided it would be best to not do that. eventually we will have the text table in the repo and itd be easily searchable

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 6]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+extern AnimationHeader D_060009A0;
+extern SkeletonHeader D_060041A8;
+extern GenericAnimationHeader D_06004404;
+
+void EnDns_Init(Actor* thisx, GlobalContext* globalCtx) {
+    EnDns* this = THIS;
+
+    if (this->actor.params < 0) {
+        osSyncPrintf("\x1b[31m引数エラー（売りナッツ）[ arg_data = %d ]\x1b[m\n", this->actor.params);
+        Actor_Kill(&this->actor);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:157]: the `\x1b[31m` chracters are VT control codes, they basically let you change the color of the text when printed to the console. youll want to include vt.h and use the macros. theres a tool in the tools folder to generate them

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 7]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+        return;
+    }
+    if (this->actor.params == 0x0006) {
+        if (LINK_AGE_IN_YEARS == YEARS_CHILD) {
+            this->actor.params = 0x0003;
+        }
+    } // Sell Seeds instead of Arrows if Link is child
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:164]: you can && these together. id also put the comment above the if instead of below

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 8]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+    }
+    if (gSaveContext.rupees < this->dnsItemEntry->itemPrice) {
+        return 0;
+    }
+    return 4;
+}
+
+u32 func_809EF9A4(EnDns* this) {
+    if (Inventory_HasEmptyBottle() == 0) {
+        return 1;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:289]: you can do !HasEmptyBottle

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 9]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+    Math_SmoothScaleMaxMinS(&this->actor.shape.rot.y, this->actor.yawTowardsLink, 3, 2000, 0);
+    this->actor.posRot.rot.y = this->actor.shape.rot.y;
+    if (func_8002F194(&this->actor, globalCtx)) {
+        this->actionFunc = EnDns_Talk;
+        return;
+    }
+    if (((this->collider.base.maskA & 2) != 0) || (this->actor.unk_10C != 0)) {
+        this->actor.flags |= 0x10000;
+    } else {
+        this->actor.flags &= 0xFFFEFFFF;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:347]: `0xFFFEFFFF` can be written as `~0x10000`

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 10]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+}
+
+void EnDns_Talk(EnDns* this, GlobalContext* globalCtx) {
+    if ((func_8010BDBC(&globalCtx->msgCtx) == 4) && (func_80106BC8(globalCtx))) {
+        switch (globalCtx->msgCtx.choiceIndex) {
+            case 0: // OK
+                switch (this->dnsItemEntry->purchaseableCheck(this)) {
+                    case 0:
+                        func_8010B720(globalCtx, 0x10A5);
+                        // "[sound 3882]Not enough Rupees!  Come back again!"
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:361]: the comments on the cases for the choice index are fine, but id remove the text id ones

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 11]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+    }
+}
+
+void func_809EFDD0(EnDns* this, GlobalContext* globalCtx) {
+    if (this->actor.params == 0x9) {
+        if (CUR_UPG_VALUE(UPG_STICKS) < 2) {
+            func_8002F434(&this->actor, globalCtx, GI_STICK_UPGRADE_20, 130.0f, 100.0f);
+            return;
+        }
+        func_8002F434(&this->actor, globalCtx, GI_STICK_UPGRADE_30, 130.0f, 100.0f);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:396]: you can remove all of these returns by using elses and else ifs

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 12]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+        this->actionFunc = func_809EFF50;
+    }
+}
+
+void func_809EFF50(EnDns* this, GlobalContext* globalCtx) {
+    if (Actor_HasParent(&this->actor, globalCtx) != 0) {
+        this->actor.parent = NULL;
+        this->actionFunc = func_809EFF98;
+        return;
+    }
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:423]: can remove != 0 and make it an else

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 13]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+    func_809EFDD0(this, globalCtx);
+}
+
+void func_809EFF98(EnDns* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+
+    if ((player->stateFlags1 & 0x400) != 0) {
+        if ((func_8010BDBC(&globalCtx->msgCtx) == 6) && (func_80106BC8(globalCtx) != 0)) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:431]: can remove != 0

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 14]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+            this->dnsItemEntry->setRupeesAndFlags(this);
+            this->dropCollectible = 1;
+            this->maintainCollider = 0;
+            this->actor.flags &= -2;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:435]: -2 is `~1`

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 15]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+            EnDns_ChangeAnim(this, 1);
+            this->actionFunc = EnDns_SetupBurrow;
+        }
+    } else {
+        this->dnsItemEntry->setRupeesAndFlags(this);
+        this->dropCollectible = 1;
+        this->maintainCollider = 0;
+        this->actor.flags &= -2;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:443]: ~

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 16]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+        EnDns_ChangeAnim(this, 1);
+        this->actionFunc = EnDns_SetupBurrow;
+    }
+}
+
+void func_809F008C(EnDns* this, GlobalContext* globalCtx) {
+    if ((func_8010BDBC(&globalCtx->msgCtx) == 6) && (func_80106BC8(globalCtx) != 0)) {
+        this->maintainCollider = 0;
+        this->actor.flags &= -2;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:452]: ~

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 17]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+        EnDns_ChangeAnim(this, 1);
+        this->actionFunc = EnDns_SetupBurrow;
+    }
+}
+
+void EnDns_SetupBurrow(EnDns* this, GlobalContext* globalCtx) {
+    f32 frameCount = SkelAnime_GetFrameCount(&D_06004404);
+
+    if (frameCount == this->skelAnime.animCurrentFrame) {
+        Audio_PlayActorSound2(&this->actor, NA_SE_EN_AKINDONUTS_HIDE);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:462]: id flip this comparison if it still matches

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 18]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+    depth = this->yInitPos - this->actor.posRot.pos.y;
+    if ((this->dustTimer & 3) == 0) {
+        initPos.x = this->actor.posRot.pos.x;
+        initPos.y = this->yInitPos;
+        initPos.z = this->actor.posRot.pos.z;
+        func_80028990(globalCtx, 20.0f, &initPos);
+    }
+    this->actor.shape.rot.y += 0x2000;
+    if (400.0f < depth) {
+        if (this->dropCollectible) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:483]: flip if it matches and && them together

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 19]
```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
...
+            initPos.x = this->actor.posRot.pos.x;
+            initPos.y = this->yInitPos;
+            initPos.z = this->actor.posRot.pos.z;
+            for (i = 0; i < 3; i++) {
+                Item_DropCollectible(globalCtx, &initPos, 3);
+            }
+        } // Drops only if you bought its item
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:490]: mentioned it before, but imo the comment at the end of the codeblock is pretty hard to follow. in general id put it right before the relevant code or beside it if fits/makes sense to

**Review by fig02 — COMMENTED** [review]
nice job on this, a few things

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c (hunk 13 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Dns/z_en_dns.c:431]: For both the stateFlags1 and the func_80106BC8?

Comment (fig02) [reply src/overlays/actors/ovl_En_Dns/z_en_dns.c:431]: i believe `func_80106BC8` returns a bool, but its not decompiled yet and not many existing examples treat it as one. its fine to leave for now i think

Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Dns/z_en_dns.c:431]: Yeah, I've seen a few like that. I'll leave it for the time being

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c (hunk 18 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Dns/z_en_dns.c:483]: Flipping works, but && no longer matches

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c (hunk 10 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Dns/z_en_dns.c:361]: I find it easier to see what NPCs with lots of dialog are doing if I can match it to the text, but I can remove them if you'd prefer

Comment (fig02) [reply src/overlays/actors/ovl_En_Dns/z_en_dns.c:361]: yeah we've decided against adding the strings into the code as it can clutter up really fast. when we have the text table in the repo it will be very easy to search

**Review by fig02 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 20]
```diff
@@ -21,63 +55,440 @@ const ActorInit En_Dns_InitVars = {
...
+extern AnimationHeader D_060009A0;
+extern SkeletonHeader D_060041A8;
+extern GenericAnimationHeader D_06004404;
+
+void EnDns_Init(Actor* thisx, GlobalContext* globalCtx) {
+    EnDns* this = THIS;
+
+    if (this->actor.params < 0) {
+        osSyncPrintf(VT_FGCOL(RED) "引数エラー（売りナッツ）[ arg_data = %d ]" VT_RST "\n", this->actor.params);
+        // FUNCTION ERROR (DEKU SALESMAN)
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:132]: We generally have translation above the print. Also avoid all caps when translating (we still have some of that in the codebase, from the old style, but they should get fixed at some point).
```suggestion
        // "Function Error (Deku Salesman)"
        osSyncPrintf(VT_FGCOL(RED) "引数エラー（売りナッツ）[ arg_data = %d ]" VT_RST "\n", this->actor.params);
```

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 21]
```diff
@@ -21,63 +55,440 @@ const ActorInit En_Dns_InitVars = {
...
+        return 1;
+    }
+    if (gSaveContext.rupees < this->dnsItemEntry->itemPrice) {
+        return 0;
+    }
+    return 4;
+}
+
+u32 func_809EF854(EnDns* this) {
+    if (CHECK_QUEST_ITEM(QUEST_GORON_RUBY) == 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:239]:
```suggestion
    if (!CHECK_QUEST_ITEM(QUEST_GORON_RUBY)) {
```

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 22]
```diff
@@ -21,63 +55,440 @@ const ActorInit En_Dns_InitVars = {
...
+    }
+}
+
+void EnDns_Wait(EnDns* this, GlobalContext* globalCtx) {
+    Math_SmoothScaleMaxMinS(&this->actor.shape.rot.y, this->actor.yawTowardsLink, 3, 2000, 0);
+    this->actor.posRot.rot.y = this->actor.shape.rot.y;
+    if (func_8002F194(&this->actor, globalCtx)) {
+        this->actionFunc = EnDns_Talk;
+    } else {
+        if (((this->collider.base.maskA & 2) != 0) || (this->actor.unk_10C != 0)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:320]:
```suggestion
        if ((this->collider.base.maskA & 2) || (this->actor.unk_10C != 0)) {
```

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 23]
```diff
@@ -21,63 +55,440 @@ const ActorInit En_Dns_InitVars = {
...
+
+void func_809EFDD0(EnDns* this, GlobalContext* globalCtx) {
+    if (this->actor.params == 0x9) {
+        if (CUR_UPG_VALUE(UPG_STICKS) < 2) {
+            func_8002F434(&this->actor, globalCtx, GI_STICK_UPGRADE_20, 130.0f, 100.0f);
+        } else {
+            func_8002F434(&this->actor, globalCtx, GI_STICK_UPGRADE_30, 130.0f, 100.0f);
+        }
+    } else {
+        if (this->actor.params == 0xA) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:370]: else if here imo

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 24]
```diff
@@ -21,63 +55,440 @@ const ActorInit En_Dns_InitVars = {
...
+        this->actionFunc = func_809EFF98;
+    } else {
+        func_809EFDD0(this, globalCtx);
+    }
+}
+
+void func_809EFF98(EnDns* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+
+    if ((player->stateFlags1 & 0x400)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:402]:
```suggestion
    if (player->stateFlags1 & 0x400) {
```

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 25]
```diff
@@ -21,63 +55,440 @@ const ActorInit En_Dns_InitVars = {
...
+    }
+    this->actor.shape.rot.y += 0x2000;
+    // Drops only if you bought its item
+    if (depth > 400.0f) {
+        if (this->dropCollectible) {
+            initPos.x = this->actor.posRot.pos.x;
+            initPos.y = this->yInitPos;
+            initPos.z = this->actor.posRot.pos.z;
+            for (i = 0; i < 3; i++) {
+                Item_DropCollectible(globalCtx, &initPos, 3);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:461]:
```suggestion
                Item_DropCollectible(globalCtx, &initPos, ITEM00_HEART);
```

**Review by Roman971 — COMMENTED** [review]
Well done with this actor. Just a few suggestions

File: src/overlays/actors/ovl_En_Dns/z_en_dns.c [hunk 26]
```diff
@@ -21,63 +55,440 @@ const ActorInit En_Dns_InitVars = {
...
+        Actor_Kill(&this->actor);
+        return;
+    }
+    // Sell Seeds instead of Arrows if Link is child
+    if ((this->actor.params == 0x0006) && (LINK_AGE_IN_YEARS == YEARS_CHILD)) {
+        this->actor.params = 0x0003;
+    }
+    osSyncPrintf(VT_FGCOL(GREEN) "◆◆◆ 売りナッツ『%s』 ◆◆◆" VT_RST "\n", D_809F0424[this->actor.params],
+                 this->actor.params);
+    // DEKU SALESMAN
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Dns/z_en_dns.c:142]:
```suggestion
    // Deku Salesman
    osSyncPrintf(VT_FGCOL(GREEN) "◆◆◆ 売りナッツ『%s』 ◆◆◆" VT_RST "\n", D_809F0424[this->actor.params],
                 this->actor.params);
```

**Review by Roman971 — APPROVED** [review]

**Review by fig02 — APPROVED** [review]
good to merge after a small undefined syms conflict is resolved
